feat(lsp): add fswatch watchfunc backend

lewis6991 opened the pull request:
### Problem

`vim._watch.poll` has terrible performance.

### Solution

On linux use fswatch as a watcher backend if available.

---

- Add File watcher section to health:vim.lsp. Warn if watchfunc is watchdirs.
- Renamed `_watch.poll` -> `_watch.watchdirs`
- All watchfuncs now support `include_pattern` and `exclude_pattern`.
- Improve type annotations all around.
- Run related tests on all watchfunc backends
- Tidied and simplified tests
- Close the file watchers when the LSP client is stopped.

Relates to: #23291

### Future work

Add a [watchmen](https://facebook.github.io/watchman/) [watchfunc](https://github.com/neovim/neovim/issues/23291#issuecomment-1712422887)

Note to reviewers: Please review by commit.

mfussenegger commented on runtime/lua/vim/lsp/health.lua:
```diff
@@ -48,4 +49,33 @@ function M.check()
+  if watchfunc == vim._watch.watch then
+    watchfunc_name = 'libuv-watch'
+  elseif watchfunc == vim._watch.poll then
+    watchfunc_name = 'libuv-poll'
```
I think we should rename the method, it's not actually polling anymore but opposed to the watch method it sets up a watcher per directory

mfussenegger reviewed (commented):
Is there any way we can have a test case to ensure this results in the expected triggers?

lewis6991 commented:
I'm happy to update any existing tests so they cover both backends.

MariaSolOs commented on runtime/lua/vim/_watch.lua:
```diff
@@ -2,15 +2,15 @@ local M = {}
+  [3] = 'Deleted',
 }
 
---- Enumeration describing the types of events watchers will emit.
```
I think we should still keep the description.

lewis6991 replied:
I dunno, it seems pretty redundant since know it's an enum, we know it's about types of events and it's defined in `_watch.lua` so must be used by watchers.

MariaSolOs replied:
Right but it's part of the "public API" of this module (and yeah the module itself is private but still).

lewis6991 replied:
Done.

MariaSolOs commented on runtime/lua/vim/_watch.lua:
```diff
@@ -198,7 +199,7 @@ end
 ---
 ---@param path (string) The path to watch. Must refer to a directory.
----@param opts (table|nil) Additional options
+---@param opts watch.PollOpts? (table|nil) Additional options
```
Isn't this redundant?
```suggestion
---@param opts watch.PollOpts? Additional options
```

lewis6991 replied:
No. it's to control the doc-gen. Which this `(table|nil)` will be documented.

MariaSolOs replied:
In that case why keep the `?`?

lewis6991 replied:
Because that is a part of the type.

lewis6991 replied:
Ah, we don't generate docs from this file so it is redundant.

MariaSolOs replied:
> Ah, we don't generate docs from this file so it is redundant.

So does this mean that you agree with me on removing the `(table|nil)`?

lewis6991 replied:
Yes, as like you said it is redundant. I assumed because the docstring was complete that we exported docs for this.

MariaSolOs replied:
Then why haven't you removed the redundancy?

MariaSolOs commented on runtime/lua/vim/_watch.lua:
```diff
@@ -221,4 +222,11 @@ function M.poll(path, opts, callback)
 
+---@param path string The path to watch. Must refer to a directory.
+---@param opts table? Additional options. Unused
+---@param callback function The function called when new events
```
```suggestion
---@param callback function The function called when new events occur.
```

MariaSolOs commented on runtime/lua/vim/_watch/fswatch.lua:
```diff
@@ -0,0 +1,77 @@
+local FSWATCH_EVENTS = {
```
I would appreciate some comments in this file :)

MariaSolOs commented on runtime/doc/news.txt:
```diff
@@ -335,6 +335,9 @@ The following changes to existing APIs or features add new behavior.
 • The `workspace/didChangeWatchedFiles` LSP client capability is now enabled
   by default.
+  • On Mac or Windows, `libuv.fs_watch` is used as the backend.
+  • On Linux, `fswatch` is used a the backend if available, otherwise
```
Should we maybe also mention here that `fswatch` is recommended?
```suggestion
  • On Linux, `fswatch` (recommended) is used a the backend if available, otherwise
```

MariaSolOs commented:
Unrelated nit but since we're already touching this file, is this function really needed? It's a simple one-liner used in a single place.

https://github.com/neovim/neovim/blob/57affc3307a0d6b3a60fe4348a8d3fde8d6f2891/runtime/lua/vim/_watch.lua#L14-L20

lewis6991 commented on runtime/lua/vim/_watch.lua:
```diff
@@ -1,101 +1,129 @@
----@return (function) Stops the watcher
+--- @param path string The path to watch
+--- @param opts vim._watch.watch.Opts? Additional options:
+---      - uvflags (table|nil)
```
I don't think this should be included and that all watchfuncs should try to support the same set of options. I've begun that here by adding universal support for `include_pattern` and `exclude_pattern`.

lewis6991 commented:
Trying to get all the backends crossed with all the platforms to agree on exact events was a bit tricky since different API calls can cause different events in any one combination.

For example `io.open` on Linux appears to cause a `Create` event with `fswatch`, however Windows with any backend requires `file:close()` to cause the `Created` event.

In practice these "event inconsistencies" are benign but does make testing a little awkward.

I've worked around it by creating files by renaming a temporary file which is how vim saves files internally anyway.

Hopefully that makes CI happy give or take some more random delays. 🤞

lewis6991 committed:
refactor(watch): general tidy up

- Rename watch.poll to watch.watchdirs
- Unify how include and exclude is applied
- Improve type hints

lewis6991 committed:
refactor(watch): simplify filechanges processing

lewis6991 committed:
feat(lsp): add fswatch watchfunc backend

Problem:
  vim._watch.watchdirs has terrible performance.

Solution:
  - On linux use fswatch as a watcher backend if available.

  - Add File watcher section to health:vim.lsp. Warn if watchfunc is
    libuv-poll.

lewis6991 committed:
fix(lsp): cancel watchers when closing a client

lewis6991 commented:
Just found we weren't cancelling the file watchers when closing the client. Added a fix here.

MariaSolOs reviewed (approved):
Looking forward to this. I've been following #23291 for a very long time :)

lewis6991 commented:
@mfussenegger I know you still want to explore a LibUV only approach, but until that is implemented can we merge this for the meantime to unblock anyone experiencing the severe performance issues.

If a viable LibUV replacement becomes available, then we can always walk this patch back.

iurimateus commented on runtime/doc/news.txt:
```diff
@@ -337,6 +337,9 @@ The following changes to existing APIs or features add new behavior.
 • The `workspace/didChangeWatchedFiles` LSP client capability is now enabled
   by default.
+  • On Mac or Windows, `libuv.fs_watch` is used as the backend.
+  • On Linux, `fswatch` (recommended) is used a the backend if available,
```
typo?
```suggestion
  • On Linux, `fswatch` (recommended) is used as the backend if available,
```

mfussenegger commented:
> @mfussenegger I know you still want to explore a LibUV only approach, but until that is implemented can we merge this for the meantime to unblock anyone experiencing the severe performance issues?
> 
> If a viable LibUV replacement becomes available, then we can always walk this patch back.

Yes fine to go ahead with this. Sorry, I didn't mean to imply that this should wait or not be done.
I think even if we have a good-enough libUV story for linux built-in, fswatch is probably still valuable to have as it allows to share watchers across processes (I think?)

I did want to give this a closer look, but on a glance nothing stood out so fine to merge. We can always follow up if something pops up.

Nice choice on the `watchdirs` naming. I wasn't sure what to call it instead.

justinmk reviewed (approved):
This is awesome and hopefully opens the door for exposing a [public file-watcher](https://github.com/neovim/neovim/issues/1380#issuecomment-546667155) interface (also for 'autoread').

Do I understand correctly that libuv seems to work well on mac/win, so we only need a fallback on Linux? Thus we wouldn't need to expose the "provider" concept (unlike e.g. clipboard), though we _internally_ have a fallback.

Does this address any of these items?: https://github.com/neovim/neovim/issues/23291#issuecomment-1817873088

> * De-duplicate debounce (`_watchfiles` has it's own mechanism)
> * Global file-path based watcher de-duplication. ...
> * Deal with watcher limits (`proc/sys/fs/inotify/max_user_watches`)
> * ✅  Configuration for excludes? ...

lewis6991 commented on runtime/lua/vim/_watch.lua:
```diff
@@ -1,186 +1,213 @@
-  assert(not stop_err, stop_err)
-  local is_closing, close_err = handle:is_closing()
-  assert(not close_err, close_err)
-  if not is_closing then
-    handle:close()
+--- An |lpeg| pattern. Only changes to files and directories whose full path does
+--- not match the pattern will be reported. Matches against both files and
+--- directories. When nil, matches nothing.
+--- @field exclude_pattern? vim.lpeg.Pattern
```
This is all private for now, but yes, I think this area needs more work.

Preferably, the LPEG thing should be an implementation detail and these fields should just be `string|string[]`.

mfussenegger commented:
>> Deal with watcher limits (proc/sys/fs/inotify/max_user_watches)
> 
> Nope, I believe this is a fundamental limit we can't work around. Again, this might improve with fanotify, but is mostly out of our control.

That bullet point was more about ensuring that the error is handled in some sensible way - e.g. aborting instead of stubbornly continue trying to setup additional handles and giving the user a sensible message - telling them to increase the limit, or inform about the limitation. I'm not sure how it behaves right now.

asmodeus812 commented:
By the way @lewis6991 worth nothing that the fs_event performance is abysmal on windows, compared to the fs_poll, however the poll i find is very often incorrect and does not report paths which are supposedly watched in comparison with fs_event. There is also this which i hope is going to be fixing the performance issues with fs_watch on windows - https://github.com/libuv/libuv/pull/4327

lewis6991 commented:
I think `fswatch` uses the same method as LibUV on Windows, so I doubt you'll see any improvements.
